### ISSUE TITLE
fzf-git functions not working on MSys2 bash (Git bash)

### ISSUE THREAD

**Issue by rashil2000** [akinomyoga/blesh-contrib#4]
Not sure if this is the right place to ask for help, but I want to give it a shot anyways.

While the functions themselves are bash statements, fzf' _preview_ argument gets executed under cmd (and not bash) on Windows. This causes certain problems, listed below:

1. Semicolon, used in the preview argument here, is not a command separator in cmd. https://github.com/akinomyoga/blesh-contrib/blob/9e145d562348d1017300a9476b401049f3f62619/fzf-git.bash#L23
2. In the preview argument here, the `<<<` operator is used which does not exist in cmd. It also uses command substitution `$(...)` which cmd doesn't support. https://github.com/akinomyoga/blesh-contrib/blob/9e145d562348d1017300a9476b401049f3f62619/fzf-git.bash#L31
3. The `<<<` operator is used here too. https://github.com/akinomyoga/blesh-contrib/blob/9e145d562348d1017300a9476b401049f3f62619/fzf-git.bash#L48

It might be too much to ask, but is there a way to modify them so that they don't use such operators?

**Comment by akinomyoga** [comment]
Thank you for the report! These functions are actually copied from https://gist.github.com/junegunn/8b572b8d4b5eddd8b85e5f4d40f17236, so maybe you also want to report it there.

----

> It might be too much to ask, but is there a way to modify them so that they don't use such operators?

You can copy the file, edit it as you like and load the edited file in `~/.blerc`.

```bash
$ mkdir -p ~/.local/share/blesh/local
$ cp /path/to/contrib/fzf-git.bash ~/.local/share/blesh/local/my-fzf-git.bash
$ EDIT ~/.local/share/blesh/local/my-fzf-git.bash
```

```bash
# blerc

ble-import -d my-fzf-git
```

The custom search path of `ble-import` can be configured by `bleopt import_path` whose default value is `${XDG_DATA_HOME:-$HOME/.local/share}/blesh/local`. The option `import_path` is a colon-separated list of directory paths just like the `PATH` environment variable.

**Comment by rashil2000** [comment]
> These functions are actually copied from https://gist.github.com/junegunn/8b572b8d4b5eddd8b85e5f4d40f17236, so maybe you also want to report it there.

Yes, I knew that but the gist interface isn't really very useful for giving feedback (and upon seeing the thread, the author doesn't seem to be very active in the comments).

Anyway, I doubt that I can edit the functions myself and make them work 😅

Thanks for quick reply. Should I close this then?

**Comment by akinomyoga** [comment]
Maybe we can work around it by specifying the shell in the following way:

```diff
-    --preview 'xxxxx' | 
+    --preview 'bash -c "xxxxx"' | 
```

but I'm not sure if that is the appropriate or intended way of `fzf`. I have checked the source of `fzf`. The preview command is used in the following code of `fzf`.

- [`junegunn/fzf` (3f90fb4) `/src/terminal.go` L1950..1952](https://github.com/junegunn/fzf/blob/3f90fb42d8871920138ace9878502f22a4d91e85/src/terminal.go#L1950-L1952)

`util.ExecCommand` is defined for Unix and Windows differently:

- [`junegunn/fzf` (3f90fb4) `/src/util/util_linux.go` L12..27](https://github.com/junegunn/fzf/blob/3f90fb42d8871920138ace9878502f22a4d91e85/src/util/util_unix.go#L12-L27)
- [`junegunn/fzf` (3f90fb4) `/src/util/util_windows.go` L13..29](https://github.com/junegunn/fzf/blob/3f90fb42d8871920138ace9878502f22a4d91e85/src/util/util_windows.go#L13-L29)

We notice that we can change the executing shell through the environment variable `SHELL` in Unix but not in Windows. Maybe we could ask for `fzf` to recognize some environment variables for the executing shell also in the Windows environment. Also, I'm wondering if it is possible to build `fzf` targeted for MSYS2 which provides the Unix API.

1. So the first thing you should do is to try to build the proper `fzf` for the MSYS2 environment using Unix API of MSYS2. If you don't know how to do that, please ask it at `fzf`.
2. If that is technically difficult, then we can next request `fzf` adding the environment variable or option to change the executing shell from `cmd` to a user-specified one.
3. If that was not accepted, finally we need to edit the preview command as `bash -c "xxxxx"` for the MSYS2 environment though I don't think it is the proper way to solve the problem.

**Comment by rashil2000** [comment]
> So the first thing you should do is to try to build the proper `fzf` for the MSYS2 environment using Unix API of MSYS2. If you don't know how to do that, please ask it at `fzf`.

I don't think this is possible. Golang only supports the native toolchain on Windows. While there are numerous Unix toolchains for Golang, they include neither Cygwin nor MSys2. The only way to do this is to build Golang itself from source by studying and defining the appropriate Go build flags (corresponding to the ones listed [here](https://golang.org/doc/install/source#environment)) and use that to build `fzf`.

Surprisingly, the Cygwin repos do include a very [old port](https://cygwin.com/cgi-bin2/package-cat.cgi?file=x86_64%2Ffzf-src%2Ffzf-0.12.1-2-src&grep=fzf) of `fzf`. Looks like it was manually patched for Cygwin. No luck for MSys2 though. The Go compiler can be installed through pacman in MSys2, but that package is in the mingw repos so I don't think it uses the Unix APIs. 

> If that is technically difficult, then you can next request `fzf` adding the environment variable or option to change the executing shell from `cmd` to a user-specified one.

This is a very tricky request. Let's say we propose a variable `$FZF_DEFAULT_SHELL`. Most people who use Windows will likely not have `bash` in their path (the bash being either from Git, MSys2 or Cygwin), so this makes the option very niche. 

Also, since the introduction of WSL, there actually is a `bash` in the path, pointing to `C:\\Windows\\System32\\bash.exe`, which is WSL's bash. We obviously don't want to accidentally use this `bash` as it will cause immense slowdowns and possibly wrong results.

> If that was not accepted, finally we need to edit the preview command as `bash -c "xxxxx"` for the MSYS2 environment though I don't think it is the proper way to solve the problem.

This might not sound good on principle, but it might just be the simplest and most non-invasive workaround. We can specify `sh -c xxxxx` only for the preview window as required, and we'll also avoid unexpected problems that can arise due to the intermingling of escape sequences if bash is used as the default execution binary for fzf on Windows.
